conformance/README.md

```diff
@@ -0,0 +1,678 @@
+When an expectation (`produces` or `signals`) is found, the preceding fragments
+are combined and then parsed and expanded.
+
+In clauses that accept fragments, more than one can be provided:
+
+```
+(in10 (text "$1")
+      (text "$2")
+      (text "$3")
+      (produces $ion $ion_1_0 $ion_symbol_table))
+```
+
+As suggested by this example, when fragments are combined, appropriate whitespace is injected
+to keep them from combining.
+
+One important constraint is that a fragment must represent zero or more complete
+top-level values, unless a `signals` clause immediately follows.
+
+The more common use of fragment combination is to allow tests to "branch out"
+from a starting point.
+
+
+It’s common in test cases to want to share a fixture or common test setup.
+The DSL allows a common document prefix to be shared across multiple test cases.
+You can use `then` clauses to continue the document in different directions,
+each with additional fragments and an expectation.
+
+````
+(in11 (text '''$ion_encoding::((module M (macro_table (macro mac [] 1)))
+                               (macro_table M))'''
+      (then (text "(:M:mac)")
+            (produces 1))
+      (then (text "(:M:0)")
+            (produces 1))
+      (then (text "(:mac)")
+            (produces 1))
+      (then (text "(:0)")
+            (produces 1)))
+````
+
+Here, each `then` clause extends the document with a different suffix,
+then checks the results.
+This more clearly associates the specific inputs forms and their outputs.
+
+A further refinement is to indicate that multiple input forms all produce the
+same output, using `each`:
+
+```
+(in11 (text '''$ion_encoding::((module M (macro_table (macro mac [] 1)))
+                               (macro_table M))'''
+      (each (text "(:M:mac)")
+            (text "(:M:0)")
+            (text "(:mac)")
+            (text "(:0)")
+            (produces 1)))
+```
+
+The `each` clause extends the document multiple directions, then verifies the
+same expectation for each result.  In this example, there are four final
+documents; all have the same `$ion_encoding` directive but each has its own
+single E-expression.
+
+This pattern makes it easy to verify multiple representations of the same data,
+ensuring that all input forms handle scenarios the same way.
+
+```
+(in11 (each
+        (text "1")
+        (bytes 0x51 0x01)
+        (toplevel 1)         // Abstract syntax representation; see below.
+        (produces 1)))
+```
+
+
+### Combining Formats
+
+TODO: What does combining text + binary mean?
+
+In theory, we can mix and match text, binary, and AST in a single
+document, since ultimately they all express the same thing in different ways.
+This requires the test-runner to not simply concatenate fragments but to
+transcribe them into one format before ultimately parsing the document and
+validating expectations.
+Of course this raises the bar on the test runner; at the very least it requires
+an Ion *writer*.
```

Comment:
Fair enough.

My overall thinking here, not well-expressed in prose, is that different implementations will approach this in different ways.  Since this is a declarative spec, that works fine, but we'll probably also need some way to ensure thorough coverage.

Reply:
I'll clean up the prose here, but will leave a more thorough design to others.